File: pr-2790/sphinx/examples/python/executing_kernels.py
# [Begin Sample]
import cudaq
import numpy as np

qubit_count = 2

# Define the simulation target.
cudaq.set_target("qpp-cpu")


# Define a quantum kernel function.
@cudaq.kernel
def kernel(qubit_count: int):
    qvector = cudaq.qvector(qubit_count)

    # 2-qubit GHZ state.
    h(qvector[0])
    for i in range(1, qubit_count):
        x.ctrl(qvector[0], qvector[i])

    # If we do not specify measurements, all qubits are measured in
    # the Z-basis by default or we can manually specify it also
    mz(qvector)


print(cudaq.draw(kernel, qubit_count))

result = cudaq.sample(kernel, qubit_count, shots_count=1000)

print(result)
# [End Sample]
''' [Begin SampleOutput]  
     ╭───╮     
q0 : ┤ h ├──●──
     ╰───╯╭─┴─╮
q1 : ─────┤ x ├
          ╰───╯

{ 11:506 00:494 }
 [End SampleOutput] '''

# [Begin `SampleAsync`]
result_async = cudaq.sample_async(kernel, qubit_count, shots_count=1000)

print(result_async.get())
# [End `SampleAsync`]
''' [Begin `SampleAsyncOutput`]
{ 00:498 11:502 }
[End `SampleAsyncOutput`] '''

# [Begin Observe]
from cudaq import spin

# Define a Hamiltonian in terms of Pauli Spin operators.
hamiltonian = spin.z(0) + spin.y(1) + spin.x(0) * spin.z(0)


@cudaq.kernel
def kernel1(n_qubits: int):
    qubits = cudaq.qvector(n_qubits)
    h(qubits[0])
    for i in range(1, n_qubits):
        x.ctrl(qubits[0], qubits[i])


# Compute the expectation value given the state prepared by the kernel.
result = cudaq.observe(kernel1, hamiltonian, qubit_count).expectation()

print('<H> =', result)
# [End Observe]
''' [Begin ObserveOutput]  
<H> = 0.0
 [End ObserveOutput] '''

# [Begin `GetState`]
# Compute the statevector of the kernel
result = cudaq.get_state(kernel, qubit_count)

print(np.array(result))
# [End `GetState`]
''' [Begin `GetStateOutput`]
[0.+0.j 0.+0.j 0.+0.j 1.+0.j]
 [End `GetStateOutput`] '''


# [Begin `ObserveAsync`]
# Define a quantum kernel function.
@cudaq.kernel
def kernel1(qubit_count: int):
    qvector = cudaq.qvector(qubit_count)

    # 2-qubit GHZ state.
    h(qvector[0])
    for i in range(1, qubit_count):
        x.ctrl(qvector[0], qvector[i])


# Measuring the expectation value of 2 different Hamiltonians in parallel
hamiltonian_1 = spin.x(0) + spin.y(1) + spin.z(0) * spin.y(1)

# Asynchronous execution on multiple `qpus` via `nvidia` `gpus`.
result_1 = cudaq.observe_async(kernel1, hamiltonian_1, qubit_count, qpu_id=0)

# Retrieve results
print(result_1.get().expectation())
# [End `ObserveAsync`]
''' [Begin `ObserveAsyncOutput`]  
1.1102230246251565e-16
 [End `ObserveAsyncOutput`] '''

File: targettests/execution/executing_kernels.cpp
/*******************************************************************************
 * Copyright (c) 2025 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// clang-format off
// Simulators
// RUN: nvq++ %cpp_std --enable-mlir  %s -o %t && %t | FileCheck %s
// RUN: nvq++ %cpp_std --library-mode %s -o %t && %t | FileCheck %s

// Quantum emulators
// RUN: nvq++ %cpp_std --target quantinuum               --emulate %s -o %t && %t | FileCheck %s
// RUN: nvq++ %cpp_std --target ionq                     --emulate %s -o %t && %t | FileCheck %s
// RUN: nvq++ %cpp_std --target anyon                    --emulate %s -o %t && %t | FileCheck %s
// 2 different IQM machines for 2 different topologies
// RUN: nvq++ %cpp_std --target iqm --iqm-machine Adonis --emulate %s -o %t && %t | FileCheck %s
// RUN: nvq++ %cpp_std --target iqm --iqm-machine Apollo --emulate %s -o %t && %t | FileCheck %s
// RUN: nvq++ %cpp_std --target oqc                      --emulate %s -o %t && %t | FileCheck %s
// RUN: if %braket_avail; then nvq++ %cpp_std --target braket --emulate %s -o %t && %t | FileCheck %s; fi
// clang-format on

#include <cudaq.h>
#include <iostream>
#include <string>
#include <vector>
#include <cstdio>
#include <cudaq/algorithms/draw.h>

using namespace std::complex_literals;
using complex = std::complex<cudaq::real>;

void kernel(int qubit_count) __qpu__ {
  cudaq::qvector qvector(qubit_count);
  // 2-qubit GHZ state.
  h(qvector[0]);
  for (auto qubit : cudaq::range(qubit_count - 1)) {
    x<cudaq::ctrl>(qvector[qubit], qvector[qubit + 1]);
  }
  // If we do not specify measurements, all qubits are measured in
  // the Z-basis by default or we can manually specify it also
//   mz(qvector);
}

void printCounts(cudaq::sample_result &result) {
    std::vector<std::string> values{};
    for (auto &&[bits, counts] : result) {
        values.push_back(bits);
    }
    
    std::sort(values.begin(), values.end());
    for (auto &&bits : values) {
        std::cout << bits << std::endl;
    }
}
    
int main() {
  int qubit_count = 2;
  {
    std::cout << "Executing kernel - sample" << std::endl;
    auto produced_str = cudaq::draw(kernel, qubit_count);
    auto counts = cudaq::sample(kernel, qubit_count);
    printCounts(counts);
  }
  // clang-format off
  // CHECK: Executing kernel - sample
  // CHECK: 00
  // CHECK: 11
  // clang-format on

  {
    std::cout << "Executing kernel - observe" << std::endl;
    // Define a Hamiltonian in terms of Pauli Spin operators.
    auto hamiltonian = cudaq::spin_op::z(0) + cudaq::spin_op::y(1) +
    cudaq::spin_op::x(0) * cudaq::spin_op::z(0);

    // Compute the expectation value given the state prepared by the kernel.
    auto counts = cudaq::observe(kernel, hamiltonian, qubit_count).expectation();
    std::cout << counts << std::endl;

    // Compute the statevector of the kernel
    cudaq::state t = cudaq::get_state(kernel, qubit_count);
  }
  // clang-format off
  // CHECK: Executing kernel - observe
  // CHECK: 1
  // clang-format on
}
